patches/tModLoader/Terraria/ModLoader/ModRarity.cs

```diff
@@ -0,0 +1,92 @@
+﻿using Microsoft.Xna.Framework;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Terraria.ID;
+
+namespace Terraria.ModLoader
+{
+	public abstract class ModRarity : ModType
+	{
+		private static int nextRarity = ItemRarityID.Count;
+
+		internal static readonly IList<ModRarity> rarities = new List<ModRarity>();
+
+		internal static int ReserveRarityID() {
+			if (ModNet.AllowVanillaClients)
+				throw new Exception("Adding item breaks vanilla client compatibility");
+			if (nextRarity == 0)
+				throw new Exception("ItemRarity ID limit has been broken");
+
+			int reserveID = nextRarity;
+			nextRarity++;
+			return reserveID;
+		}
+
+		/// <summary>
+		/// Returns the ModRarity associated with specified type
+		/// If not a ModRarity, returns null.
+		/// </summary>
+		/// <param name="type"></param>
+		/// <returns></returns>
+		public static ModRarity GetRarity(int type) {
+			return type >= ItemRarityID.Count && type < RarityCount ? rarities[type - ItemRarityID.Count] : null;
+		}
+
+		public static int RarityCount => nextRarity;
+
+		internal static void Unload() {
+			rarities.Clear();
+			nextRarity = ItemRarityID.Count;
+		}
+
+		public int Type { get; internal set; }
+
+		protected override void Register() {
+			if (!Mod.loading)
+				throw new Exception("AddRarity can only be called from Mod.Load or Mod.Autoload");
+
+			if (Mod.prefixes.ContainsKey(Name))
+				throw new Exception($"You tried to add 2 ModRarities with the same name: {Name}. Maybe 2 classes share a classname but in different namespaces while autoloading or you manually called AddRarity with 2 rarities of the same name.");
+
+			Type = ReserveRarityID();
+
+			Mod.rarities[Name] = this;
+			rarities.Add(this);
+			ContentInstance.Register(this);
+		}
+
+		/// <summary>
+		/// Your ModRarity's color.
+		/// Returns White by default.
+		/// </summary>
+		/// <returns></returns>
+		public virtual Color RarityColor() => Color.White;
+
+		/// <summary>
+		/// Your modded rarity plus one. Used for prefixes. Returns your modded rarity by default.
+		/// </summary>
+		/// <returns></returns>
+		public virtual int RarityPlusOne() => GetRarity(Type).Type;
```

Comment:
That's all pretty hardcoded, and needs a redesign. I understand the need for ordering now.

Reply:
What is this  even? And why are there so many? Shouldn't this just be an overrideable property?